### ISSUE TITLE
172-huawei-e3372-macos

### ISSUE THREAD

**Issue by retifrav** [retifrav/retifrav.github.io#1]
[Huawei E3372 modem on Mac OS](https://retifrav.github.io/blog/2019/04/17/huawei-e3372-macos/) comments.

**Comment by retifrav** [comment]
ololo

**Comment by retifrav** [comment]
Some new comment with:

- *italic*
- **bold**
- `code`

**Comment by zs-stpa** [comment]
does no longer work in macOS 10.15 catalina. any ideas there?

**Comment by retifrav** [comment]
Haven't installed it yet, so I cannot check. But if it's because of the driver, then the only way would be to get an updated driver, which hopefully will be released by vendor (Huawei) at some point.

**Comment by mattsson** [comment]
How do you send `AT` commands using the Terminal and `screen`? I feel like I'm pretty well-versed in using the Terminal, but I'm having a hard time finding any information on how to do this.

**Comment by retifrav** [comment]
Like it's said in the article, if the modem is connected and recognized in the system (`ls -l /dev/tty.HUA*`), then you just launch `screen /dev/tty.HUAWEIMobile-Pcui` (or whatever ID it has in your device tree) and then you can execute `AT` commands.

**Comment by mattsson** [comment]
I do have the screen session running, but how do I send commands? Simply entering the letters `AT` and pressing return in the Terminal does nothing.

**Comment by retifrav** [comment]
I don't remember having issues with that, but let me check it once I'm home. I'll let you know.

**Comment by mattsson** [comment]
Ah, never mind. The command doesn't appear when I write it put it does actually get input. So, writing `AT` does give me an `OK`. However, inputting `AT^SETPORT="A1;10,12,16"` just gives me an `ERROR` back.

**Comment by retifrav** [comment]
@zs-stpa, if you are still looking for Catalina driver, [here it is](https://retifrav.github.io/blog/2019/04/17/huawei-e3372-macos/drivers/catalina/MobileConnectDriver.pkg).

**Comment by mattsson** [comment]
@retifrav Awesome, thank you!

**Comment by kate-is-great** [comment]
I have also arrived in this dark place (Catalina). Directly after installing the device works (e.g., internet connection, can check it via 192.168.8.1, shows up in network connections). However, if I unplug and plug again it does not connect anymore. 

/Library/Extensions/HuaweiDataCardDriver_10_9.kext is in place. 

Right after the installation (when it works) it cannot find the device with $ ls -l /dev/tty.HUA* and hence I cannot complete the steps thereafter described above. 

Any ideas? Have spent 8+ hours with Huawei Support, Apple Support...

**Comment by retifrav** [comment]
That sounds weird, it wouldn't be able to connect if driver wasn't working. Unplugging and plugging pack should not affect anything. I don't know what could be wrong here. Perhaps, try good-old reboot, if you haven't yet rebooted since installing the driver.

**Comment by kate-is-great** [comment]
Thanks for your quick feedback. It is weird. I did reboot and even cleared NVRAM and SMC. No success. 

Is the linked version 6.00.06.00 the latest version of this or is there a newer one? 

It is really amazing how this can be so damn hard. There must be 100 m of these things in use and a decent share with Catalina...

**Comment by retifrav** [comment]
These two drivers that I've uploaded here (*one x32 for pre-Catalina and another x64 for Catalina*) are the only ones I've got, I don't know if they had more updates since then. But given the Mac share among their users (*I think it's super small*), Huawei just doesn't care that much, I guess.

**Comment by DnyaneshwarWagh** [comment]
**For macOS Big Sur (11.0.1) Refer**

[https://consumer.huawei.com/en/support/content/en-us15773565/
](https://consumer.huawei.com/en/support/content/en-us15773565/
)

Instead of disabling master SIP you can disable partial SIP by using

`csrutil enable --without kext`

**Comment by retifrav** [comment]
Thank you for the information! Looks like the driver is the same (*`MobileConnectDriver.pkg` from `Mobile Partner.app` contents, the size is equal*), but apparently on Big Sur it now requires disabling SIP (*or at least `kext`*). Sadly, I couldn't test it in my VM (*Parallels refuses to boot it in Recovery mode*), so I'll have to wait till I get a physical machine with Big Sur.

**Comment by DnyaneshwarWagh** [comment]
I have tested Huawei E3276 on MacBook Pro with macOS Big Sur (11.1) and it is working.
For Parallels recovery mode this might help you.
[ https://kb.parallels.com/cn/116526](https://kb.parallels.com/cn/116526)

**Comment by retifrav** [comment]
Yeah, saw that one already, and a couple of other KBs and forum posts - nothing works: it either boots to regular mode or never boots at all. But it's good that you can confirm that modem and driver work fine on your machine with Big Sur.

**Comment by joscelyn** [comment]
Thanks for the sharing ! Tried with my E5372 on macOS Big Sur (11.1) by using the DnyaneshwarWagh link: kexts are installed but nothing in `/dev/` and nothing in "Network"

**Comment by DnyaneshwarWagh** [comment]
> Thanks for the sharing ! Tried with my E5372 on macOS Big Sur (11.1) by using the DnyaneshwarWagh link: kexts are installed but nothing in `/dev/` and nothing in "Network"

1. Disable SIP as explained in 'Driver Installation Guide for MAC 11_2.docx'
2. After Every System Reboot, need to load kext by using following command.
`sudo kextload /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/*.kext`

**Comment by joscelyn** [comment]
Thanks DnyaneshwarWagh.
Even with that, nothing appends, it only prints:
`Executing: /usr/bin/kmutil load -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardDMM.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiLogLevel.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBActivateDriver.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBEthernetData.kext`

**Comment by DnyaneshwarWagh** [comment]
> Thanks DnyaneshwarWagh.
> Even with that, nothing appends, it only prints:
> `Executing: /usr/bin/kmutil load -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardDMM.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiLogLevel.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBActivateDriver.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBEthernetData.kext`
1. Can you post output of `csrutil status` command.
2. After `sudo kextload /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/*.kext` command, reconnect E5372, wait few seconds then execute `ls -l /dev/tty.HUA*`

**Comment by joscelyn** [comment]
It works ! I needed to unplug the modem, thanks DnyaneshwarWagh :)

**Comment by driversat** [comment]
works great on BigSur 11.1 hackintosh with opencore 0.6.4 after adding HuaweiDataCardDriver_10_9.kext to EFI/OC/Kext and injecting it to config.plist with opencore-configurator

**Comment by OneWhiteSpirit** [comment]
Is it work on arm64 with Rosetta on BigSur?

**Comment by BIGmakus1980** [comment]
Hi, someone fixed the problems with this router after the recent Mac BigSur updates (10 / 11.2021)? My version is 11.2.3.

**Comment by retifrav** [comment]
Mac OS 11.2.3 is from March 2021, so it's not very recent. I currently have 11.6.1, and the modem is working fine there.

**Comment by BIGmakus1980** [comment]
> Mac OS 11.2.3 is from March 2021, so it's not very recent. I currently have 11.6.1, and the modem is working fine there.

Have you made any changes to SIP settings etc?

**Comment by retifrav** [comment]
Yes, and I described it in the post, you might have missed the [latest update](https://retifrav.github.io/blog/2019/04/17/huawei-e3372-macos/#28102021) from 28.10.2021.

**Comment by BIGmakus1980** [comment]
Thanks, I upgraded to 11.6.1 and did what I needed to do, but still something doesn't work...


Last login: Fri Nov 26 14:23:35 on ttys000
piotrsobkowiak@MacBook-Pro-Piotr ~ % ls -L1 /dev/tty.HUA*
/dev/tty.HUAWEIMobile-Diag
/dev/tty.HUAWEIMobile-Modem
/dev/tty.HUAWEIMobile-Pcui
piotrsobkowiak@MacBook-Pro-Piotr ~ % csrutil status
System Integrity Protection status: unknown (Custom Configuration).

Configuration:
	Apple Internal: disabled
	Kext Signing: disabled
	Filesystem Protections: enabled
	Debugging Restrictions: enabled
	DTrace Restrictions: enabled
	NVRAM Protections: enabled
	BaseSystem Verification: enabled

This is an unsupported configuration, likely to break in the future and leave your machine in an unknown state.
piotrsobkowiak@MacBook-Pro-Piotr ~ % sudo kextload /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/*.kext
Executing: /usr/bin/kmutil load -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardDMM.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiLogLevel.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBActivateDriver.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBEthernetData.kext
Password:
Executing: /usr/bin/kmutil load -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardDMM.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMControl.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiDataCardECMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/HuaweiLogLevel.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBACMData.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBActivateDriver.kext -p /Library/Extensions/HuaweiDataCardDriver_10_9.kext/Contents/PlugIns/MBBEthernetData.kext
zsh: command not found: Executing:
piotrsobkowiak@MacBook-Pro-Piotr ~ %

**Comment by retifrav** [comment]
Can't say what could be wrong here, the only thing I can guess is that you seem to be running Zsh, while I am still using Bash, so I would suggest you to run this from Bash as well. Might be a silly idea, but that's the only guess I got.

Also just in case check that `kmutil` is available:

```
/usr/bin/kmutil --help
```

**Comment by BIGmakus1980** [comment]
piotrsobkowiak@MacBook-Pro-Piotr ~ % /usr/bin/kmutil --help
OVERVIEW: kmutil: KernelManagement Utility (KernelManagement_executables-102.120.15)

USAGE: kmutil <subcommand>

OPTIONS:
  -h, --help              Show help information.

SUBCOMMANDS:
  create                  Create one or more new artifacts based on the arguments provided.
  load                    Load one or more extensions based on the arguments provided.
  unload                  Unload the named kexts and all personalities.
  log                     Display logging information about the KernelManagement subsystem.
  libraries               Search for library kexts that define symbols needed for linking by a kernel extension.
  dumpstate               Dumps kernelmanagerd(8) state for debugging
  inspect                 Inspect & display a kext collection's contents according to the options provided.
  clear-staging           Clears all contents of the kext staging locations on the system
  configure-boot          Configure custom boot object policy (recovery-mode only)
  find                    Find kexts available on the operating system.
  showloaded              Show the loaded state of the extensions on the system, according to the options provided.
  trigger-panic-medic     Delete and disable loading of third party kexts in order to safely boot into a target volume. (can only be triggered in Recovery mode)
                          eg usage: `kmutil trigger-panic-medic --volume-root /Volumes/<VolumeName>`
  check                   Check the consistency of kext collections against each other and/or load information in-kernel.
  print-diagnostics       Perform all possible tests on a specified kext, and indicate whether the kext is loadable.

  See 'kmutil help <subcommand>' for detailed help.
